lib/rls/api.rb

```diff
@@ -122,19 +123,61 @@ def seasons(renew = false)
       end
     end
 
+    # Performs a mutex-protected request to the API with rate limit handling.
     # @param type [String, Symbol] HTTP verb
     # @param endpoint [String, Symbol] The API endpoint
     # @param attributes [Array<Hash>] Header and query parameters
     #   passed along with the request
     # @return [Hash] The parsed JSON response
     def request(type, endpoint, *attributes)
       attributes << {} if attributes.empty?
-      response = raw_request(type, endpoint, attributes)
-      JSON.parse(response)
+      @mutex ||= Mutex.new
+
+      @mutex.synchronize do
+        sleep until_reset if will_be_rate_limited?
+
+        begin
+          @last_response = raw_request(type, endpoint, attributes)
+        rescue RestClient::TooManyRequests
+          sleep until_reset
+          retry
+        end
+
+        JSON.parse(@last_response)
+      end
     end
 
     private
 
+    # @return [String, nil] if it exists, the specific header from the last API request
+    def last_header(key)
+      @last_response&.headers&.dig(key)
+    end
+
+    # @return [Float] the amount of time until the rate limit resets
+    def until_reset
+      (last_header(:x_rate_limit_reset_remaining) || 0).to_i / 1000.0
+    end
+
+    # @return [Integer] number of requests until we're rate limited
+    def remaining_requests
+      last_header(:x_rate_limit_remaining)&.to_i || -1
```

Comment:
In case this is confusing:

`last_header(:x_rate_limit_remaining)&.to_i` returns `[Integer, nil]` which would fail the subsequent `#zero?` check. I opted just to return `-1` as it isn't zero.